### PR TITLE
CURL_PUSH_ERROROUT: allow the push callback to fail the parent stream

### PR DESCRIPTION
**Pull request by bagder**
... by adding support for a new dedicated return code.

Suggested-by: Jonathan Cardoso
URL: https://curl.haxx.se/mail/lib-2020-06/0099.html

### CODE REVIEW

File: docs/libcurl/opts/CURLMOPT_PUSHFUNCTION.3 [hunk 1]
```diff
@@ -86,6 +86,9 @@ the ownership of the CURL handle has been taken over by the application.
 .IP "CURL_PUSH_DENY (1)"
 The callback denies the stream and no data for this will reach the
 application, the easy handle will be destroyed by libcurl.
+.IP "CURL_PUSH_ERROROUT (2)"
+Returning this will reject the pushed stream and return an error back on the
+parent stream making it get closed with an errror. (Added in curl 7.72.0)
```
Comment (JDepooter) [inline docs/libcurl/opts/CURLMOPT_PUSHFUNCTION.3:91]: Small typo here.

Comment (bagder) [reply docs/libcurl/opts/CURLMOPT_PUSHFUNCTION.3:91]: thanks!

File: lib/http2.c [hunk 2]
```diff
@@ -563,6 +565,7 @@ static int push_promise(struct Curl_easy *data,
     stream->push_headers_used = 0;
 
     if(rv) {
+      DEBUGASSERT((rv > CURL_PUSH_OK) && (rv < CURL_PUSH_ERROROUT));
```
Comment (erijo) [inline lib/http2.c:568]: Shouldn't it be `<=` instead of `<`?

Comment (bagder) [reply lib/http2.c:568]: Yes it should indeed, thanks!

File: lib/http2.c [hunk 3]
```diff
@@ -737,11 +741,16 @@ static int on_frame_recv(nghttp2_session *session, const nghttp2_frame *frame,
   case NGHTTP2_PUSH_PROMISE:
     rv = push_promise(data_s, conn, &frame->push_promise);
     if(rv) { /* deny! */
-      rv = nghttp2_submit_rst_stream(session, NGHTTP2_FLAG_NONE,
+      int h2;
+      DEBUGASSERT((rv > CURL_PUSH_OK) && (rv < CURL_PUSH_ERROROUT));
```
Comment (erijo) [inline lib/http2.c:745]: Same here, shouldn't this be <= as well?

Comment (bagder) [reply lib/http2.c]: Thanks for your keen eyes, yes!

**Comment by bagder** [comment]
Thanks for your help @erijo!

**Pull request closed without merging**